### PR TITLE
Fix is_plugin_added::<Self>() being true during build

### LINKED ISSUE #13815: is_plugin_added is always true for Self
## Bevy version

0.14.0-rc.2

## What you did

The following program panics in Bevy 0.14.0-rc.2, but runs successfully in 0.13.2

```rust
use bevy::prelude::*;

fn main() {
    let app = App::new().add_plugins(Foo);
}

pub struct Foo;

impl Plugin for Foo {
    fn build(&self, app: &mut App) {
        if app.is_plugin_added::<Self>() {
            panic!()
        }
    }
}
```

## What went wrong

Either the previous behavior should be restored, or we should document it in the migration guide.

### PR DESCRIPTION
**Pull request by kristoff3r**
# Objective

Fixes #13815 

## Solution

Move insertion of the plugin name to after build is called.

## Testing

I added a regression test

### CODE REVIEW

File: crates/bevy_app/src/app.rs [hunk 1]
```diff
@@ -1274,4 +1272,20 @@ mod tests {
             .init_non_send_resource::<NonSendTestResource>()
             .init_resource::<TestResource>();
     }
+
+    #[test]
+    /// Plugin should not be considered inserted while it's being built
+    ///
+    /// bug: <https://github.com/bevyengine/bevy/issues/13815>
+    fn regression_test_13815() {
```
Comment (alice-i-cecile) [inline crates/bevy_app/src/app.rs:1280]:
```suggestion
    fn plugin_should_not_be_added_during_build_time() {
```

**Review by alice-i-cecile — APPROVED** [review]
I like this behavioral change, and the code is clearer too. I'd like to change the test name (I find the links more useful in the comments than the name) but that's just a nit.

**Comment by BD103** [comment]
Should this be considered a breaking change?

**Comment by alice-i-cecile** [comment]
IMO no: this restores us to 0.13 behavior.

**Review by mockersf — APPROVED** [review]

File: crates/bevy_app/src/app.rs [hunk 2]
```diff
@@ -1275,6 +1273,21 @@ mod tests {
...
+        pub struct Foo;
+
+        impl Plugin for Foo {
+            fn build(&self, app: &mut App) {
+                assert!(!app.is_plugin_added::<Self>());
+            }
+        }
+
+        App::new().add_plugins(Foo);
+
```
Comment (mockersf) [inline crates/bevy_app/src/app.rs:1290]:
```suggestion
        App::new().add_plugins(Foo);
    }
```

Comment (mockersf) [reply crates/bevy_app/src/app.rs:1290]: messed up conflict resolution, sorry...

**Pull request merged**